Local calendar integration

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Local calendar integration using an rfc5545 local calendar and full support for recurring events. Extends the entity model with new mutation methods invoked by a websocket for supporting add and delete from the UI (update is possible, just pruned for frontend simplicity for now).

See https://github.com/allenporter/ical/blob/main/DESIGN.md#recurrence for details about how recurring events are implemented in the underlying library, basically:
- rrules are exposed to the UI as a string for editing. The frontend UI can render recurrence rule UIs, but serialize as an rfc5545 string like `FREQ=DAILY;COUNT=3` etc.
- recurring events are implemented as a single event when written to disk
- however, they are expanded on the fly and presented to home assistant as flattened so home assistant doesn't have to care about the (this is similar to how google calendar handles recurring events)
- recurring events are identified by a `uid` for the whole series and a `recurrence_id` for an individual instance
- The `recurrence_range` can support distinguishing between "delete this event" and "delete this and all future events"


Frontend PR: https://github.com/home-assistant/frontend/pull/14036
Architecture Discussion: https://github.com/home-assistant/architecture/discussions/797
ical library documentation: https://allenporter.github.io/ical/ical.html
ical library changelogs: https://github.com/allenporter/ical/releases

Planned work in this PR:
- [X] Local event storage
- [X] Add/Edit websockets
- [X] Recurring event support
- [X] "supports" feature flags so frontend knows which calendars are mutable
- [X] More testing of timezones (need to either never use local time, or teach the calendar library about home assistants local timezone)
- [ ] Tests for websocket and demo calendar, including "not implemented" case
- [x] Related Dev API PR
- [ ] End user docs PR

Future work in other PRs:
- implement for google calendar
- improve performance of the store?
- update support

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- 
- Link to documentation pull request: 
  - https://github.com/home-assistant/home-assistant.io/pull/24429
  - https://github.com/home-assistant/developers.home-assistant/pull/1488

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Add local calendar integration

Add rrule support for creating events

Update local calendar version and service selector

Add local calendar websocket for add, update, cancel

Update ical version

Add end to end test of recurring events

Rename cancel method to delete

allenporter committed:
Move local calendar websocket to calendar integration

homeassistant commented:
Hey there @home-assistant/core, mind taking a look at this pull request as it has been labeled with an integration (`calendar`) you are listed as a [code owner](https://github.com/home-assistant/core/blob/dev/CODEOWNERS#L174) for? Thanks!

allenporter commented:
@MartinHjelmare while there are a couple loose ends i'd like to get to before this is fully ready for merging, i think there is enough here for an initial review to make sure the key up front decisions on design are right. Can you take a look when you have time?

allenporter committed:
Set timezone in tests as a temporary workaround

allenporter committed:
Update tests for new fields

allenporter committed:
Use Home Assistant timezone when converting all day events to dates

allenporter committed:
Update fields returned in tests

allenporter committed:
Add calendar supported features

allenporter committed:
Add uniqueids and bump ical fix rrule serialization fix

allenporter commented:
The frontend PR had an initial pass and is now marked `wait for backend`.

allenporter commented:
I just saw there is a timezone handling bug and a test is failing, oops. I'll take a look, though expect its a minor fix compared to the overall PR, or may be a library fix.

MartinHjelmare commented on homeassistant/components/local_calendar/calendar.py:
```diff
@@ -0,0 +1,154 @@
+    calendar = IcsCalendarStream.calendar_from_ics(ics)
+
+    name = config_entry.data[CONF_CALENDAR_NAME]
+    entity_id = generate_entity_id(ENTITY_ID_FORMAT, name, hass=hass)
```
Why do we need to set an entity_id? Why not just use the entity name as a default template for the entity_id?

MartinHjelmare commented on homeassistant/components/local_calendar/calendar.py:
```diff
@@ -0,0 +1,154 @@
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Get all events in a specific time frame."""
+        with use_local_timezone(dt_util.DEFAULT_TIME_ZONE):
```
What does this context manager do?

allenporter replied:
From https://allenporter.github.io/ical/ical/util.html#use_local_timezone
> Set the local timezone to use when converting a date to datetime.
> This is expected to be used as a context manager when the default timezone used by python is not the timezone to be used for calendar operations (the attendees local timezone).

Example:
```
import datetime
import zoneinfo
from [ical.calendar](https://allenporter.github.io/ical/ical/calendar.html) import Calendar
from [ical.event](https://allenporter.github.io/ical/ical/event.html) import Event
from [ical.util](https://allenporter.github.io/ical/ical/util.html) import use_local_timezone

cal = Calendar()
cal.events.append(
    Event(
        summary="Example",
        start=datetime.date(2022, 2, 1),
        end=datetime.date(2022, 2, 2)
    )
)
# Use UTC-8 as local timezone
with use_local_timezone(zoneinfo.ZoneInfo("America/Los_Angeles")):
    # Returns event above
    events = cal.timeline.start_after(
        datetime.datetime(2022, 2, 1, 7, 59, 59, tzinfo=datetime.timezone.utc))

    # Does not return event above
    events = cal.timeline.start_after(
        datetime.datetime(2022, 2, 1, 8, 00, 00, tzinfo=datetime.timezone.utc))
```

My impression is that home assistant's timezone is not the same thing as the local time used by python. This API is trying to solve the problem of passing timezone information all the way down into a python library when the timezone comparison is being made. 

An alternative can be to pass it into the `timeline` object and have it do what it wants with this timezone from the caller. I'm happy to flex on how this works.  Also, this is where the current bug in CI is that i need to dig in and investigate.

MartinHjelmare replied:
I'd require the date times passed by the caller to be timezone aware and just go with those.

allenporter replied:
Done.

MartinHjelmare commented on tests/components/caldav/test_calendar.py:
```diff
@@ -939,5 +939,8 @@ async def test_get_events_custom_calendars(hass, calendar, get_api_events):
             "summary": "This is a normal event",
             "location": "Hamburg",
             "description": "Surprisingly rainy",
+            "uid": None,
```
Why do we update other calendar integration tests?

allenporter replied:
Two things happening here:
1) This PR changes the entity model that any integration can use. These new fields are in use by local calendar, but will also be used by google calendar.
2) The current behavior of the rest API is to return a field (e.g. `description`) even when it is not set. my impression is that some APIs prefer this so that callers don't have to handle multiple conditions of unset and set to null, but I don't have a preference here.  PR https://github.com/home-assistant/core/pull/79598 codifies this behavior more clearly showing the differences between the API return values and the other `as_dict` method used in event payloads.

From your comment, i take it that you'd like #2 to omit `None` fields. Am I understanding that right?

allenporter committed:
Move calendar fields inline websocket

Move calendar fields inline websocket, mark some fields as required,
and remove some other optional fields that are used for update only.

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -35,6 +52,19 @@
 # mypy: disallow-any-generics
 
 
+CALENDAR_EVENT_SCHEMA = vol.Schema(
```
Done. I realized some of the fields were here for `update` which I removed from this pr, so those have also been removed.

allenporter committed:
Use generic entity component

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -310,3 +363,96 @@ async def get(self, request: web.Request) -> web.Response:
+        raise HomeAssistantError("Calendar integration not set up")
+
+    if (entity := component.get_entity(entity_id)) is None or not isinstance(
+        entity, CalendarEntity
```
I see there is now a generic EntityComponent also (or there always was and I didn't know about it). Removed.

allenporter committed:
Replace error codes with existing constants

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -310,3 +363,96 @@ async def get(self, request: web.Request) -> web.Response:
+    try:
+        entity = _get_calendar_entity(hass, msg["entity_id"])
+    except HomeAssistantError as ex:
+        connection.send_error(msg["id"], "failed", str(ex))
```
Fixed.

allenporter committed:
Remove unnecessary entity id logic

allenporter committed:
Fix typing on hass websocket fixture

allenporter commented on tests/components/local_calendar/test_calendar.py:
```diff
@@ -0,0 +1,515 @@
+        return resp.get("result")
+
+
+ClientFixture = Callable[[], Client]
```
Fixed.

allenporter commented on homeassistant/components/local_calendar/calendar.py:
```diff
@@ -0,0 +1,154 @@
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import generate_entity_id
```
This was removed based on other comment.

allenporter commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -35,6 +52,19 @@
+        vol.Optional(EVENT_END): vol.Any(cv.date, cv.datetime),
+        vol.Optional(EVENT_SUMMARY): cv.string,
+        vol.Optional(EVENT_DESCRIPTION): cv.string,
+        vol.Optional(EVENT_RRULE): cv.string,
```
The RRULE spec is defined here:
https://www.rfc-editor.org/rfc/rfc5545#section-3.3.10

The rrule value is currently validated in the integration library here: https://github.com/allenporter/ical/blob/main/ical/types/recur.py

However, what we can also do is add more validation here of the form:
- verifying that it parses as an a rrule at all
- verifying specific parts of rrule home assistant wants to support and enforce across all integrations

I think there are some parts of rrule we don't want to or need to support (e.g. recurrence rules that repeat more often than daily), so we could enforce that here rather than in the integration.  I think validating the `FREQ` field could be a reasonable thing to do here, but maybe not enforcing constraints on all the other rule parts.

If we ever support event editing, one challenge is that you might imagine that an integration could support an rrule that the frontend doesn't know how to create. We can either pass those through unvalidated, but probably could be simplest to just not support updates.

So I lean towards using `datetutil.rrule` for a "light" validation of `FREQ` here.  Seem OK?

MartinHjelmare replied:
Sounds good.

allenporter replied:
Added.

allenporter commented:
Thanks for the thorough review, you hit on the right points.

allenporter committed:
Simplify event fetching with local timezone

allenporter committed:
Add light rrule validation

MartinHjelmare commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -230,6 +283,19 @@ async def async_get_events(
         """Return calendar events within a datetime range."""
         raise NotImplementedError()
 
+    async def async_create_event(self, **kwargs: Any) -> dict[str, Any]:
```
If we only need the event uid from the return value, why should we return a dict?

allenporter replied:
OK, reduced scope.

With regard to the websocket value, really, the uid is only used in tests at the moment and not in the frontend -- but tests could be made a little more complicated to query to get the uid.  I think it's probably Ok to leave, but just wanted to be fully transparent.

MartinHjelmare replied:
I'd consider removing it, if we don't really need it.

allenporter committed:
Reduce scope of event uid

allenporter committed:
Reduce scope of return value to what is needed

allenporter commented:
I fixed one other issue related to iteration order of all day events so they are now correctly returned in the order based on the local timezone (API added in ical 4.0.0).

MartinHjelmare reviewed (approved):
Looks good!

Can ge merged when the frontend PR is approved.

allenporter committed:
Remove unnecessary uid return value

allenporter committed:
Separate CREATE_EVENT and DELETE_EVENT into separate features
